server/hedley/modules/custom/hedley_reports/hedley_reports.module

```diff
@@ -5674,193 +5680,3072 @@ function hedley_reports_generate_completion_data_for_tuberculosis($participant,
+          break;
+        }
+      }
+
+      // If section was shown, check if patient was referred to program.
+      // Note the that check is at current encounter.
+      if ($section_expected) {
+        $speciality_care_measurement = $encounter_data['measurements_by_type']['prenatal_speciality_care'];
+        if (!empty($speciality_care_measurement) && !empty($speciality_care_measurement->field_speciality_care_signs[LANGUAGE_NONE])) {
+          foreach ($speciality_care_measurement->field_speciality_care_signs[LANGUAGE_NONE] as $sign) {
+            if ($sign['value'] == 'arv') {
+              $activity_expected = TRUE;
+              break;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Checking referral to NCD program.
+  if (!$activity_expected) {
+    $section_expected = FALSE;
+    $previous_nurse_encounters_data = $encounter_data['previous_nurse_encounters_data'];
+    foreach ($previous_nurse_encounters_data as $previous_encounter_data) {
+      // If not known as HIV positive, check by diagnosis.
+      $triggering_diagnoses = array_merge(
+        hedley_reports_prenatal_hypertensionlike_diagnoses(),
+        hedley_reports_prenatal_diabetes_diagnoses()
+      );
+      if (hedley_reports_prenatal_diagnosed_any_of($previous_encounter_data, $triggering_diagnoses)) {
+        $section_expected = TRUE;
+        break;
+      }
+    }
+
+    // If section was shown, check if patient was referred to program.
+    // Note the that check is at current encounter.
+    if ($section_expected) {
+      $speciality_care_measurement = $encounter_data['measurements_by_type']['prenatal_speciality_care'];
+      if (!empty($speciality_care_measurement) && !empty($speciality_care_measurement->field_speciality_care_signs[LANGUAGE_NONE])) {
+        foreach ($speciality_care_measurement->field_speciality_care_signs[LANGUAGE_NONE] as $sign) {
+          if ($sign['value'] == 'ncd') {
+            $activity_expected = TRUE;
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  // Checking referral to Ultrasound.
+  if (!$activity_expected) {
+    $lmp_measurement = $encounter_data['measurements_by_type']['last_menstrual_period'];
+    if (!empty($lmp_measurement)) {
+      $confident = $lmp_measurement->field_confident[LANGUAGE_NONE][0]['value'];
+      if (!$confident) {
+        if (!empty($lmp_measurement->field_not_confident_reason[LANGUAGE_NONE][0]['value'])) {
+          $activity_expected = TRUE;
+        }
+      }
+    }
+  }
+
+  if ($activity_expected) {
+    $expected[] = 'prenatal_send_to_hc';
+  }
+}
+
+/**
+ * Adds nurse Health Education activity, if conditions match.
+ *
+ * @param array $encounter_data
+ *   An associative array containing data about the encounter.
+ * @param array &$expected
+ *   An array of expected activity names, passed by reference.
+ */
+function hedley_reports_prenatal_add_health_education_activity_for_nurse(array $encounter_data, array &$expected) {
```

Comment:
```
// Instead of one large function
function hedley_reports_prenatal_add_health_education_activity_for_nurse() {
    // Split into smaller functions:
    hedley_reports_check_hiv_education();
    hedley_reports_check_partner_education();
    hedley_reports_check_diagnosis_education();
    // etc.
}
```

Perhaps if we have time for it. This is also a hard-to-read function.

Reply:
Well, I feel we've spent to much time on this feature already, so, maybe we can do it latter on in a follow up.